server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractRunnable.java

```diff
@@ -20,6 +20,13 @@ public boolean isForceExecution() {
         return false;
     }
 
+    /**
+     * Should the runnable start a new tracing context before it executes?
+     */
+    public boolean useNewTraceContext() {
```

Comment:
Maybe it makes more sense to just create a utility method on `ContextPreservingAbstractRunnable` that wraps an `AbstractRunnable` in a `ContextPreservingAbstractRunnable` with new trace context `true` flag set to `true` and use that?
That way we don't have to leak the high level tracing into low level `AbstractRunnable`. It's also probably less verbose than overriding to return `true` in multiple spots.

Reply:
I’m struggling to see how this would work. `ContextPreservingAbstractRunnable` is private and only created in `ThreadContext.preserveContext(Runnable),` but I need to communicate the need for a new trace context from where the runnable is created in the first place, for example in `InboundHandler`.

I thought about creating a named sub-class of `AbstractRunnable` e.g. `TracingAbstractRunnable` or something like that, but I couldn't see how that would avoid the need to changing `AbstractRunnable`, at least not without duplicating much of `ContextPreservingAbstractRunnable`.

I don't think there's a simpler answer for where to insert this logic, because there's a lot of parts that are collaborating here. In the case of `InboundHandler`, we see:

   * `InboundHandler` creates an `AbstractRunnable` instance and schedules it
   * The scheduler asks the thread context to wrap the runnable
   * The thread context potentially wraps the runnable
   * The scheduler executes the runnable
   * `AbstractRunnable#run` orchestrates `doRun`, `onFailure` etc

We could change the visibility of `ContextPreservingAbstractRunnable` maybe, and sometimes create one explicitly? e.g. `ContextPreservingAbstractRunnable.preservingContext(AbstractRunnable)`?

Reply:
I think you could just add a method to `ThreadContext` similar to `org.elasticsearch.common.util.concurrent.ThreadContext#preserveContext`  that takes an `AbstractRunnable` and does the wrapping for you without even having to expose `ContextPreservingAbstractRunnable`? 

You have the thread context available via the `threadPool` and the threadpool will just pass through a `ContextPreservingAbstractRunnable` unchanged when executing => less code in the callsites and no need to change anything about `AbstractRunnable` right?

Reply:
@original-brownbear ah, you're right. I had another go, please take another look.